### DISCUSSION TITLE
TerminusDB with Kubernetes ?

### DISCUSSION THREAD

**Discussion by hpvd** [terminusdb/terminusdb#1872] [Ideas]
terminusDB offers already many possibilities/features an enterprise graphdb for certain use cases should offer.

From my perspective, there is one missing,
which gives a big hurdle to use it.

Most enterprises today rely on Kubernetes or will do in the near future, because
- they really need it today
- they prepare for the future, or simply because
- it's something one "should use" today ; -)

=> so all of them, when looking for
a. a graph database or
b. an enterprise database or
c. an...what ever... database

search for `a/b/c + Kubernetes`.

Searching this way, they will hardly find/discover terminusDB.

Even if they find it, and search on
- terminusdb.com
- github.com/orgs/terminusdb/discussions

they are not welcomed with open arms, because there are zero results, see

![2023-06-26_11h35_34](https://github.com/terminusdb/terminusdb/assets/5681880/b82c60db-626a-4e5f-984a-5410a339ca3e)

![2023-06-26_11h37_03](https://github.com/terminusdb/terminusdb/assets/5681880/b7d520f3-3534-4add-87bf-b8d8be5615f6)

**Comment by hpvd** [comment]
if one look to other databases, there are some really great approaches/examples how one could make discoverability and usage very easy.
e.g. for Postgre see
- https://github.com/cloudnative-pg/cloudnative-pg
- https://cloudnative-pg.io/

**Comment by hpvd** [comment]
some great background on databases within kubernetes (-> howto) is shown in this video from devopstoolkit:
https://www.youtube.com/watch?v=Ny9RxM6H6Hg
(they also pick the example from above)

**Comment by rrooij** [comment]
You are raising a very good concern. There should indeed be more documentation and resources on how to deploy TerminusDB on Kubernetes.

Fortunately I can say that it is possible to use TerminusDB on a Kubernetes cluster. We have TerminusDB deployed ourselves in a Kubernetes cluster on AWS. It would be good to document how we approached this. There are multiple solutions: either you have one deployment with a local disk, which is less scalable. Or you have multiple deployments with a shared storage like NFS, which is better scalable but sacrifices read/write performance. In the future, we would like to implement mechanisms to increase the scalability even further while sacrificing less performance, like using blob storage but with a fast local cache.

**Comment by hpvd** [comment]
great to here you already have a running system in Kubernetes!
(for database handling in Kubernetes, the video mentioned above is really worth investing 19min...)

**Comment by hpvd** [comment]
As idea, just the list of features, what is possible/made accessible with this "cloudnative postgre" kubernetes operator:
https://cloudnative-pg.io/documentation/1.20/#main-features

**Comment by hpvd** [comment]
as addition to cloudnative-pg database example,
this is a perfect intro to Custom Resource Definitions (CRD) and Custom Controller (CC)
https://thenewstack.io/kubernetes-crds-what-they-are-and-why-they-are-useful/

and this is another db specific topic which can be solved in kubernetes with CRD and CC:
schema management/migrations:
language-neutral, database schema ORM as kubernetes Operator
yes it's about sql, but it's about the methods...
different methods/needs/solution: https://www.youtube.com/watch?v=zkmtignxbv4
https://github.com/schemahero/schemahero

CRD+CC=>Operator
In Kubernetes language, an operator is a piece of software that maps operational knowledge (hence the name) about the operation and installation of a particular program or system component in turn into software and can thus automate it.
official doc kubernetes Operator Pattern:
https://kubernetes.io/docs/concepts/extend-kubernetes/operator/
